### PR TITLE
[FLINK-30211] Introduce CompactorSink for compact jobs in Table Store

### PR DESCRIPTION
**Pull request by tsreaper**
This ticket introduces the `CompactorStoreSink` for compact jobs in Table Store.

The behavior of `CompactorStoreSink` is sketched as follows: This sink accepts records containing partitions and buckets to compact and perform compaction on these buckets.

### CODE REVIEW

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/CompactorStreamPartitioner.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * {@link StreamPartitioner} for stand-alone compact job sinks so that the same bucket is compacted
+ * by the same sink parallelism.
+ */
+public class CompactorStreamPartitioner extends StreamPartitioner<RowData> {
+
+    private final RowType rowType;
+
+    private transient RowDataSerializer serializer;
+
+    public CompactorStreamPartitioner(RowType partitionType) {
```
Comment (JingsongLi) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/CompactorStreamPartitioner.java:44]: I think we can have a `HashRowStreamPartitioner`, here pass a row type to here.
The row type can be get from `BucketsTable.rowType(RowType partitionType)`.

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/AbstractFlinkSink.java [hunk 2]
```diff
@@ -158,4 +125,10 @@ private void assertCheckpointConfiguration(StreamExecutionEnvironment env) {
                         + ExecutionCheckpointingOptions.CHECKPOINTING_MODE.key()
                         + " to exactly-once");
     }
+
+    protected abstract OneInputStreamOperator<RowData, Committable> createWriteOperator(
```
Comment (JingsongLi) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/AbstractFlinkSink.java:129]: `createWriteOperator(StoreSinkWrite.Provider writeProvider, boolean isStreaming);`

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/AbstractFlinkSink.java [hunk 3]
```diff
@@ -158,4 +125,10 @@ private void assertCheckpointConfiguration(StreamExecutionEnvironment env) {
...
+            String initialCommitUser, boolean isStreaming);
+
+    protected abstract OneInputStreamOperator<Committable, Committable> createCommitterOperator(
```
Comment (JingsongLi) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/AbstractFlinkSink.java:132]: Here we can just a `CommitRetry`, let subclasses only return different ones and mask the same part.

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/AbstractFlinkSink.java [hunk 4]
```diff
@@ -33,75 +33,45 @@
...
-import javax.annotation.Nullable;
-
 import java.io.Serializable;
-import java.util.Map;
 import java.util.UUID;
 
-/** Sink of dynamic store. */
-public class StoreSink implements Serializable {
+/** Abstract sink of table store. */
+public abstract class AbstractFlinkSink implements Serializable {
```
Comment (JingsongLi) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/AbstractFlinkSink.java:43]: Maybe just `FlinkSink`, consistent with `FlinkSource`.

**Review by JingsongLi — COMMENTED** [review]
Thanks @tsreaper for the contribution, left comments.

**Review by JingsongLi — APPROVED** [review]
+1

**Comment by Aitozi** [comment]
Hi @tsreaper @JingsongLi , I have a question here. Why we have to use AtMostOnce semantics for the compact job here ? Why we not just do the same as the writer job ? Maybe the lost COMPACT commit may lead to the data loss ?

**Pull request merged**